spec/filters/integration/multi_stage_spec.rb

```diff
@@ -1,71 +1,43 @@
 # encoding: utf-8
 require "logstash/devutils/rspec/spec_helper"
-require "logstash/filters/mutate"
 
-# running mutate which depends on grok  outside a logstash package means
-# LOGSTASH_HOME will not be defined, so let's set it here
-# before requiring the grok filter
-unless LogStash::Environment.const_defined?(:LOGSTASH_HOME)
-  LogStash::Environment::LOGSTASH_HOME = File.expand_path("../../../", __FILE__)
-end
-
-describe LogStash::Filters::Mutate do
-  let(:pipeline) do
-    new_pipeline_from_string(config)
-  end
-
-  let(:events) do
-    arr = event.is_a?(Array) ? event : [event]
-    arr.map do |evt|
-      LogStash::Event.new(evt.is_a?(String) ? LogStash::Json.load(evt) : evt)
-    end
+module LogStash::Environment
+  # running mutate which depends on grok  outside a logstash package means
+  # LOGSTASH_HOME will not be defined, so let's set it here
+  # before requiring the grok filter
+  unless self.const_defined?(:LOGSTASH_HOME)
+    LOGSTASH_HOME = File.expand_path("../../../", __FILE__)
   end
 
-  let(:results) do
-    pipeline.instance_eval { @filters.each(&:register) }
-    results  = []
-    events.each do |evt|
-      # filter call the block on all filtered events, included new events added by the filter
-      pipeline.filter(evt) do |filtered_event|
-        results.push(filtered_event)
-      end
+  # also :pattern_path method must exist (due grok filter)
+  unless self.method_defined?(:pattern_path)
+    def pattern_path(path)
+      ::File.join(LOGSTASH_HOME, "patterns", path)
     end
-    pipeline.flush_filters(:final => true) { |flushed_event| results << flushed_event }
-
-    results.select { |e| !e.cancelled? }
   end
+end
 
-  describe 'MUTATE-33: multi stage with json, grok and mutate, Case mutation' do
-    let(:event) do
-      "{\"message\":\"hello WORLD\",\"lower1\":\"PPQQRRSS\",\"lower2\":\"pppqqq\"}"
-    end
+describe 'LogStash::Filters::Mutate' do
 
+  context 'MUTATE-33: multi stage with json, grok and mutate, Case mutation' do
     let(:config) do
       <<-CONFIG
-filter {
-    grok {
+    filter {
```

Comment:
This may be too much work for the initial intent of this PR, but most of these filters have no need to use pipelines for their testing, we could just instantiate the plugin, so that the test only has to define the plugin settings and the input event:

```ruby
let(:plugin) { described_class.new(settings) }
let(:settings) {  { "lowercase" => [ "field1" ] } }
let(:event) { LogStash::Event.new(input) }
let(:input) { { } }

before(:each) do
  plugin.register
  # or even
  plugin.filter(event)
end

describe "#lowercase" do  
  let(:settings) {  { "lowercase" => [ "field1" ] } }
  let(:input) { { "field1" => "HELLLLOOO" }} 
  it "lowercases the specified field" do
    expect(event.get("field1")).to eq("hellllooo")
  end
end
```

Reply:
thanks, think the intent here was a kind of an "integration" test (grok plugin also used).
now I did not investigate why its there to start with - probably testing things together : 
```
"MUTATE-33: multi stage with json, grok and mutate, Case mutation"
```
thus since this isn't a plain-old unit spec we rather would keep the functional test, wdyt?
(also fine with dropping/migrating the "integration" spec if it does not seem worthwhile)

Reply:
My comment was more on the overall move from using pipeline + sample + insist to plugin instantiation + filter + expect.

For this particular spec, I think we should keep it until there's a replacement for it. looking at #33 and its fix it's possible to replicate this without grok, but out of scope here.

For now let's merge as is, good job :)

Reply:
cool thanks - appreciate the tip regarding not using sample/config :+1: 
delivered a unit spec rewrite in the grok plugin, takes a lot of :hourglass_flowing_sand: we need a LS specific AI!